api/capi/src/nnstreamer-capi-single-new.c

```diff
@@ -34,18 +34,124 @@
+ * @note unittests take over 5sec to complete
  */
-#define SINGLE_DEFAULT_TIMEOUT 3000
+#define SINGLE_DEFAULT_TIMEOUT 10000
+
+/** convert time in millisecond to timespec format */
+#define MSEC_TO_TIMESPEC(ts, msec) do { \
+  (ts).tv_sec = (msec) / 1000; \
+  (ts).tv_nsec = ((msec) % 1000) * 1000000; \
+} while (0)
 
 /* ML single api data structure for handle */
 typedef struct
 {
-  GTensorFilterSingle *filter;
-  ml_tensors_info_s in_info;
-  ml_tensors_info_s out_info;
+  GTensorFilterSingle *filter;  /**< tensor filter element */
+  ml_tensors_info_s in_info;    /**< info about input */
+  ml_tensors_info_s out_info;   /**< info about output */
+
+  pthread_t thread;             /**< thread for invoking */
+  pthread_mutex_t mutex;        /**< mutex for synchronization */
+  pthread_cond_t cond;          /**< condition for synchronization */
+  ml_tensors_data_h input;      /**< input received from user */
+  ml_tensors_data_h * output;   /**< output to be sent back to user */
+  struct timespec timeout;      /**< timeout for invoking */
+  gboolean data_ready;          /**< data is ready to be processed */
+  gboolean join;                /**< thread should be joined */
+  int status;                   /**< status of processing */
 } ml_single;
 
+/**
+ * @brief thread to execute calls to invoke
+ */
+static void *
+invoke_thread (void * arg)
+{
+  ml_single *single_h;
+  GTensorFilterSingleClass *klass;
+  GstTensorMemory in_tensors[NNS_TENSOR_SIZE_LIMIT];
+  GstTensorMemory out_tensors[NNS_TENSOR_SIZE_LIMIT];
+  ml_tensors_data_s *in_data, *out_data;
+  int i, status = ML_ERROR_NONE;
+
+  single_h = (ml_single *) arg;
+
+  pthread_mutex_lock (&single_h->mutex);
+
+  /** get the tensor_filter element */
+  klass = g_type_class_peek (G_TYPE_TENSOR_FILTER_SINGLE);
+  if (!klass) {
+    single_h->join = TRUE;
+    goto exit;
+  }
+
+  while (single_h->join != TRUE) {
+
+    /** wait for data */
+    while (single_h->data_ready != TRUE) {
+      pthread_cond_wait (&single_h->cond, &single_h->mutex);
+      if (single_h->join == TRUE)
+        goto exit;
+    }
+
+    in_data = (ml_tensors_data_s *) single_h->input;
+
+    /** Setup input buffer */
+    for (i = 0; i < in_data->num_tensors; i++) {
+      in_tensors[i].data = in_data->tensors[i].tensor;
+      in_tensors[i].size = in_data->tensors[i].size;
+      in_tensors[i].type = single_h->in_info.info[i].type;
+    }
+
+    /** Setup output buffer */
+    for (i = 0; i < single_h->out_info.num_tensors; i++) {
+      /** memory will be allocated by tensor_filter_single */
+      out_tensors[i].data = NULL;
+      out_tensors[i].size = ml_tensor_info_get_size (&single_h->out_info.info[i]);
+      out_tensors[i].type = single_h->out_info.info[i].type;
+    }
+    pthread_mutex_unlock (&single_h->mutex);
+
+    /** invoke the thread */
+    if (klass->invoke (single_h->filter, in_tensors, out_tensors) == FALSE) {
+      status = ML_ERROR_INVALID_PARAMETER;
+      goto wait_for_next;
```

Comment:
It will broadcast at `wait_for_next` in the situation that a mutex is unlocked.

Reply:
updated.